src/libraries/System.Private.CoreLib/src/System/DateOnly.cs

```diff
@@ -0,0 +1,902 @@
+        /// The DateOnly object will be formatted in short form.
+        /// </summary>
+        /// <returns>A string that contains the short date string representation of the current DateOnly object.</returns>
+        public override string ToString() => ToString("d");
+
+        /// <summary>
+        /// Converts the value of the current DateOnly object to its equivalent string representation using the specified format and the formatting conventions of the current culture.
+        /// </summary>
+        /// <param name="format">A standard or custom date format string.</param>
+        /// <returns>A string representation of value of the current DateOnly object as specified by format.</returns>
+        public string ToString(string? format) => ToString(format, DateTimeFormatInfo.CurrentInfo);
+
+        /// <summary>
+        /// Converts the value of the current DateOnly object to its equivalent string representation using the specified culture-specific format information.
+        /// </summary>
+        /// <param name="provider">An object that supplies culture-specific formatting information.</param>
+        /// <returns>A string representation of value of the current DateOnly object as specified by provider.</returns>
+        public string ToString(IFormatProvider? provider) => ToString("d", provider);
+
+        /// <summary>
+        /// Converts the value of the current DateOnly object to its equivalent string representation using the specified culture-specific format information.
+        /// </summary>
+        /// <param name="format">A standard or custom date format string.</param>
+        /// <param name="provider">An object that supplies culture-specific formatting information.</param>
+        /// <returns>A string representation of value of the current DateOnly object as specified by format and provider.</returns>
+        public string ToString(string? format, IFormatProvider? provider)
+        {
+            if (format == null || format.Length == 0)
+            {
+                format = "d";
+            }
+
+            if (format.Length == 1)
+            {
+                switch (format[0])
+                {
+                    case 'o':
+                    case 'O':
+                        {
+                            Span<char> destination = stackalloc char[10];
+                            bool b = DateTimeFormat.TryFormatDateOnlyO(Year, Month, Day, destination);
+                            Debug.Assert(b);
+
+                            return destination.ToString();
+                        }
+
+                    case 'r':
+                    case 'R':
+                        {
+                            Span<char> destination = stackalloc char[16];
+                            bool b = DateTimeFormat.TryFormatDateOnlyR(DayOfWeek, Year, Month, Day, destination);
```

Comment:
We could stackalloc `destination` once on top and slice here. I'd expect better code generation/optimization in the case.

Reply:
I want to allocate the exact size to ensure failure if we have any bug. I am not sure even either that will generate better optimized code. anyway I am not seeing this critical for perf.